### PR TITLE
refactor(grid): change schar_T representation to be more compact

### PR DESCRIPTION
**Pull request by bfredl**
Previously, a screen cell would occupy 28+4=32 bytes per cell as we always made space for up to MAX_MCO+1 codepoints in a cell.

As an example, even a pretty modest 50*80 screen would consume

    50*80*2*32 = 256000, i e a quarter megabyte

With the factor of two due to the TUI side buffer, and even more when using msg_grid and/or ext_multigrid

This instead stores a 4-byte union of either:
- a valid UTF-8 sequence up to 4 bytes
- an escape char which is invalid UTF-8 (0xFF) plus a 24-bit index to a glyph cache

This avoids allocating space for huge composed glyphs _upfront_, while still keeping rendering such glyphs reasonably fast (1 hash table lookup + one plain index lookup). If the same large glyphs are using repeatedly on the screen, this is still a net reduction of memory/cache consumption. The only case which really gets worse is if you blister the screen full with crazy emojis and zalgo text and even this case only leads to 4 extra bytes per char.

When only <= 4-byte glyphs are used, plus the 4-byte attribute code, i e 8 bytes in total there is a factor of four reduction of memory use. Memory which will be quite hot in cache as the screen buffer is scanned over in win_line() buffer text drawing

### CODE REVIEW

File: src/nvim/drawscreen.c [hunk 1]
```diff
@@ -444,6 +444,15 @@ int update_screen(void)
   display_tick++;  // let syntax code know we're in a next round of
                    // display updating
 
+  // glyph cache full, very rare
+  if (schar_cache_clear_if_full()) {
+    // must use CLEAR, as the contents of screen buffers cannot be
+    // compared to their previous state here.
+    // TODO(bfredl): if start to cache schar_T values in places (like fcs/lcs)
+    // we need to relavidate these here as well!
```
Comment (zeertzjq) [inline src/nvim/drawscreen.c:452]:
```suggestion
    // we need to revalidate these here as well!
```

File: src/nvim/map_glyph_cache.c [hunk 2]
```diff
@@ -0,0 +1,94 @@
+#include <nvim/api/private/helpers.h>
+#include <nvim/map.h>
```
Comment (zeertzjq) [inline src/nvim/map_glyph_cache.c:2]:
```suggestion
#include "nvim/api/private/helpers.h"
#include "nvim/map.h"
```
This file also doesn't have a PVS header.

File: src/nvim/mouse.c [hunk 3]
```diff
@@ -1904,7 +1904,7 @@ static void mouse_check_grid(colnr_T *vcolp, int *flagsp)
 
       // Remember the character under the mouse, might be one of foldclose or
       // foldopen fillchars in the fold column.
-      mouse_char = utf_ptr2char((char *)gp->chars[off]);
+      mouse_char = schar_get_ascii(gp->chars[off]);
```
Comment (zeertzjq) [inline src/nvim/mouse.c:1907]: This doesn't look correct. There can be a multibyte char in foldcolumn.

File: src/nvim/tui/tui.c [hunk 4]
```diff
@@ -703,7 +705,7 @@ static bool cheap_to_print(TUIData *tui, int row, int col, int next)
         return false;
       }
     }
-    if (strlen(cell->data) > 1) {
+    if (schar_get_ascii(cell->data) > 1) {
```
Comment (zeertzjq) [inline src/nvim/tui/tui.c:708]: This also doesn't seem to make sense.

File: src/nvim/mouse.c (hunk 3 above)
Comment (bfredl) [reply src/nvim/mouse.c:1907]: right, foldclose/foldopen can be multibyte as well. but looks weird and unreliable to deduce this back from physical grid text, and not the actual state whether this winline had an open or closed fold.

File: src/nvim/mouse.c [hunk 5]
```diff
@@ -1901,22 +1899,12 @@ static void mouse_check_grid(colnr_T *vcolp, int *flagsp)
...
-
-    // Check for position outside of the fold column.
-    if (wp->w_p_rl ? click_col < wp->w_width_inner - fdc :
-        click_col >= fdc + (cmdwin_type == 0 ? 0 : 1)) {
-      mouse_char = ' ';
     }
   }
 
-  if (wp && mouse_char == wp->w_p_fcs_chars.foldclosed) {
+  if (wp && col_from_screen == -2) {
```
Comment (zeertzjq) [inline src/nvim/mouse.c:1905]:
```suggestion
  if (col_from_screen == -2) {
```

File: src/nvim/drawline.c [hunk 6]
```diff
@@ -106,6 +106,8 @@ typedef struct {
   int c_extra;               ///< extra chars, all the same
   int c_final;               ///< final char, mandatory if set
 
+  colnr_T n_closing;         ///< number of chars in fdc which will be closing
```
Comment (zeertzjq) [inline src/nvim/drawline.c:109]: I think this should just be an `int`
```suggestion
  int n_closing;             ///< number of chars in fdc which will be closing
```

File: src/nvim/drawline.c [hunk 7]
```diff
@@ -405,7 +408,7 @@ static void handle_foldcolumn(win_T *wp, winlinevars_T *wlv)
 ///
 /// Assume monocell characters
 /// @return number of chars added to \param p
-size_t fill_foldcolumn(char *p, win_T *wp, foldinfo_T foldinfo, linenr_T lnum)
+size_t fill_foldcolumn(char *p, win_T *wp, foldinfo_T foldinfo, linenr_T lnum, colnr_T *n_closing)
```
Comment (zeertzjq) [inline src/nvim/drawline.c:411]:
```suggestion
size_t fill_foldcolumn(char *p, win_T *wp, foldinfo_T foldinfo, linenr_T lnum, int *n_closing)
```

File: src/nvim/drawline.c [hunk 8]
```diff
@@ -447,16 +450,23 @@ size_t fill_foldcolumn(char *p, win_T *wp, foldinfo_T foldinfo, linenr_T lnum)
     }
   }
 
+  colnr_T n_closing_val = i;
```
Comment (zeertzjq) [inline src/nvim/drawline.c:453]:
```suggestion
  int n_closing_val = i;
```

File: src/nvim/grid.c [hunk 9]
```diff
@@ -56,25 +61,122 @@ void grid_adjust(ScreenGrid **grid, int *row_off, int *col_off)
...
+
+schar_T schar_from_str(char *str)
+{
+  if (str == NULL) {
+    return 0;
+  }
+  return schar_from_buf(str, strlen(str));
+}
+
+// buf need not be NUL terminated, but `buf` may not contain embedded NULS.
```
Comment (zeertzjq) [inline src/nvim/grid.c:86]:
```suggestion
// buf need not be NUL terminated, but may not contain embedded NULs.
```

File: src/nvim/grid.c [hunk 10]
```diff
@@ -56,25 +61,122 @@ void grid_adjust(ScreenGrid **grid, int *row_off, int *col_off)
...
+
+/// @return ascii char or NUL if not ascii
+char schar_get_ascii(schar_T sc)
+{
+#ifdef ORDER_BIG_ENDIAN
+  if (!(sc & 0x80FFFFFF)) {
+    return *(char *)sc;
+  }
+#else
+  if (sc < 127) {
```
Comment (zeertzjq) [inline src/nvim/grid.c:167]:
```suggestion
  if (sc < 128) {
```

File: src/nvim/grid.h [hunk 11]
```diff
@@ -33,30 +33,18 @@ EXTERN colnr_T *linebuf_vcol INIT(= NULL);
 // screen grid.
 
 /// Put a ASCII character in a screen cell.
-static inline void schar_from_ascii(char *p, const char c)
-{
-  p[0] = c;
-  p[1] = 0;
-}
+#define schar_from_ascii(x) ((schar_T)(x))
```
Comment (zeertzjq) [inline src/nvim/grid.h:36]:
```suggestion
#ifdef ORDER_BIG_ENDIAN
# define schar_from_ascii(x) ((schar_T)(x << 24))
#else
# define schar_from_ascii(x) ((schar_T)(x))
#endif
```

File: src/nvim/grid.c [hunk 12]
```diff
@@ -554,9 +657,9 @@ void grid_put_linebuf(ScreenGrid *grid, int row, int coloff, int endcol, int cle
       if (line_off2cells(linebuf_char, off, max_off_from) > 1) {
         // When the first half of a double-width character is
         // overwritten, change the second half to a space.
-        schar_from_ascii(linebuf_char[off + 1], ' ');
+        linebuf_char[off + 1] = ' ';
```
Comment (zeertzjq) [inline src/nvim/grid.c:660]:
```suggestion
        linebuf_char[off + 1] = schar_from_ascii(' ');
```

File: src/nvim/grid.c [hunk 13]
```diff
@@ -554,9 +657,9 @@ void grid_put_linebuf(ScreenGrid *grid, int row, int coloff, int endcol, int cle
...
       }
-      schar_from_ascii(linebuf_char[off], '<');
+      linebuf_char[off] = '<';
```
Comment (zeertzjq) [inline src/nvim/grid.c:662]:
```suggestion
      linebuf_char[off] = schar_from_ascii('<');
```

**Comment by bfredl** [comment]
https://github.com/neovim/neovim/pull/25217#issuecomment-1723154236 using this approach I could test this on a virtualized big endian envirionment. I ran the `test/functional/ui` tests to start with and they all pass on this branch (except for a misdetected 'termguicolors' which fails also on master).

**Pull request merged**